Rpi 4.9.y ad5758 support

stefpopa opened the pull request:
The AD5758 is a single channel DAC with 16-bit precision which uses the SPI
interface that operates at clock rates up to 50MHz.

The output can be configured as voltage or current and is available on a
single terminal.

Datasheet:
http://www.analog.com/media/en/technical-documentation/data-sheets/ad5758.pdf

commodo commented on include/dt-bindings/iio/adi,ad5758.h:
```diff
@@ -0,0 +1,54 @@
+#define _DT_BINDINGS_ADI_AD5758_H
+
+
+#define	AD5758_DCDC_MODE_POWER_OFF	0
```
nitpick; indentation is inconsistent;
first part uses tabs [between `#define` and constant name], second part uses spaces;

commodo commented on Documentation/devicetree/bindings/iio/dac/ad5758.txt:
```diff
@@ -0,0 +1,97 @@
+		   can be found in <dt-bindings/iio/adi,ad5758.h>.
+
+		   The following values are currently supported:
+		   	* AD5758_DCDC_ILIM_150_mA
```
i wonder if defining these constant names is a bit much; i mean, you could easily just use values 150, 200, 250, etc [direclty] and document here that it's mili-amps

commodo commented on Documentation/devicetree/bindings/iio/dac/ad5758.txt:
```diff
@@ -0,0 +1,97 @@
+		* AD5758_SR_CLOCK_256_HZ
+		* AD5758_SR_CLOCK_128_HZ
+		* AD5758_SR_CLOCK_64_HZ
+		* AD5758_SR_CLOCK_16_HZ
```
same for these values and and `slew rate step`

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,725 @@
+#define AD5758_DIG_DIAG_RES_ERR_3WI_MSK			BIT(12)
+#define AD5758_DIG_DIAG_RES_RES_OCCURRED_MSK		BIT(13)
+#define AD5758_DIG_DIAG_RES_SLEW_BUSY_MSK		BIT(14)
+#define AD5758_DIG_DIAG_RES_CAL_MEM_UNREFRESHED_MSK	BIT(15)
```
if upstreaming this driver later, it's likely that you will have to remove these #defines [since they're not yet used] ;
if you plan to use them later, it's fine to leave them

commodo commented on Documentation/devicetree/bindings/iio/dac/ad5758.txt:
```diff
@@ -0,0 +1,97 @@
+		adi,dc-dc-mode = <AD5758_DCDC_MODE_DPC_VOLTAGE>;
+		adi,dc-dc-ilim = <AD5758_DCDC_ILIM_200_mA>;
+		adi,slew = <1 AD5758_SR_CLOCK_200_KHZ AD5758_SR_STEP_12_LSB>;
+		adi,range = <AD5758_RANGE_0V_10V>;
```
you could use an array of 2 values here; first value being minimum value, second being maximum value; and document valid values; and then you can search them in a table;
if values are invalid, you could print an err message;

it's less verbose like that and does the same thing;

stefpopa replied:
There is no signed data type for property values in device tree source files. 
I could get rid of the defines and do it like in this example: https://github.com/analogdevicesinc/linux/blob/master/Documentation/devicetree/bindings/iio/dac/ad5755.txt#L54

commodo replied:
I also read up quickly about signed values in DTs ; and I agree this is a tricky one.

I'm not feeling strongly about this [i.e. dropping the #defines and just documenting the values].
Either way works for me, unless there are other opinions.

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,725 @@
+	[AD5758_RANGE_MINUS_1mA_PLUS_22mA] = { -1, 22 },
+};
+
+static void ad5758_get_min_max(struct ad5758_state *st,
```
curios: will this function be re-used later ?
if not, then i would just add these 2 lines in the place where this is called
if yes, i would make it inline

also [curios] do `int *min, int *max)` fit on the first line ?

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,725 @@
+	return spi_write(st->spi, &st->data[0].d8[0], 4);
+}
+
+static int ad5758_spi_write_mask(struct ad5758_state *st,
```
unrelated to this PR, it does feel like this construct [of write_mask] is repeated quite a bit in IIO at least
it may be an interesting idea to wrap it somehow later;

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,725 @@
+	 */
+	timeout = 4;
+	do {
+		mdelay(1);
```
this always waits at least 1 mdelay() ;
curios: is it possible that the transaction could finish a bit earlier ?
in that case it may make sense to read the reg, check the mask, and break the loop [or return 0] earlier ;
and wait only 1 mdelay() [or more] if the value is not what we expect

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,725 @@
+	do {
+		mdelay(1);
+		readval = ad5758_spi_reg_read(st, AD5758_DCDC_CONFIG2);
+	} while (--timeout && (readval & AD5758_DCDC_CONFIG2_BUSY_3WI_MSK));
```
same comment here as https://github.com/analogdevicesinc/linux/pull/95/commits/a0e4a30c65232b8b6dbd0a6cc5b42a4e94556c68#r188181697

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,725 @@
+	return 0;
+}
+
+static int ad5758_wait_for_busy_3wi(struct ad5758_state *st)
```
you could create a common function for ` ad5758_wait_for_busy_3wi()` and `ad5758_wait_for_refresh_cycle()` and wrap it with these 2 functions; the logic is common, except for the reg & mask;

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,725 @@
+		return ret;
+
+	/* Perform a software reset */
+	ret = ad5758_soft_reset(st);
```
maybe we could also add a DT entry for a reset GPIO/pin;
and if we have a reset pin, we can call that, and then alternatively do a soft reset [if no GPIO]

but that can come in a later commit/stage

stefpopa replied:
I will add GPIO support in a later commit

commodo replied:
cool
thanks :)

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,725 @@
+
+	ret = ad5758_spi_reg_write(st, AD5758_KEY, AD5758_KEY_CODE_RESET_2);
+
+	mdelay(1);
```
according to the docs, you can check the `AD5758_DIG_DIAG_RES_RES_OCCURRED_MSK` bit to check if a reset occurred;

stefpopa replied:
There are a few issues with this approach which might overly complicate things. 
The RESET_OCCURRED flag is set to 1 at power up, so first thing that needs to be done is to clear it. Furthermore, a software reset initializes all registers to their default values, therefore CRC will be enabled. As for now, CRC is not supported by this driver.

commodo replied:
ok
let's leave it like this then; i'm not feeling strongly about this;
thanks :)

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,725 @@
+		ret = ad5758_spi_reg_write(st, AD5758_DAC_INPUT, val);
+		break;
+	default:
+		ret = -EINVAL;
```
i wonder if the compiler would complain here if there's a missing `break` statement;
GCC 7+ has some new defaults that complain a bit more than older versions;

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,725 @@
+
+	if (device_property_read_u32(&st->spi->dev, "adi,dc-dc-ilim",
+				     &dc_dc_ilim))
+		dev_warn(&st->spi->dev,
```
i would not raise this to WARN log level; it would spam dmesg ; maybe leave it to DEBUG level

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,725 @@
+	if (ret < 0)
+		return ret;
+
+	if (device_property_read_u32(&st->spi->dev, "adi,dc-dc-mode",
```
what if you group all these `device_property_read_xxx()` functions into a single call ?
and you wrap it into a `ad5758_parse_dt()` function or something ?

then you could initialize all default values there, and call each setter with the value from the state;

example: `ad5758_set_dc_dc_conv_mode(st, st->dc_dc_mode);`  ( given that `st->dc_dc_mode` has been init-ed from DT or has a default value)

lclausen-adi commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,725 @@
+
+	indio_dev = devm_iio_device_alloc(&spi->dev, sizeof(*st));
+	if (indio_dev == NULL)
+		return  -ENOMEM;
```
extra space

lclausen-adi commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,725 @@
+	int ret;
+
+	indio_dev = devm_iio_device_alloc(&spi->dev, sizeof(*st));
+	if (indio_dev == NULL)
```
I think there is a cocci script that will replace this with !indio_dev

lclausen-adi commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,725 @@
+
+	st->pwr_down = pwr_down;
+
+	return ret ? ret : len;
```
ret is always 0 at this point.

lclausen-adi commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,725 @@
+	unsigned int dac_config_msk, dac_config_mode;
+	int ret;
+
+	ret = kstrtou32(buf, 10, &pwr_down);
```
strtobool() maybe

stefpopa commented:
changelog v1 -> v2:
* removed the unused defines
* completely removed the adi,ad5758.h file
* there are no more constant names in ad5758.txt; values are used instead
* removed the `ad5758_get_min_max()` function as it was used only once
* created a new, more generic ad5758_wait_for_task_complete() function which waits until CAL_MEM_UNREFRESHED or BUSY_3WI are 0. Also, this function first reads the reg value,  checks the mask and returns if the value is the one that is expected, otherwise waits.
* included a break statement for the default case from `ad5758_write_raw()`
* did a `!!st->pwr_down` instead of a bool cast inside the `ad5758_read_powerdown()` function
* created a new `ad5758_parse_dt()` function
* used `dev_dbg `instead of `dev_warn `for the case when the properties are missing from the devicetree
* removed the extra space from return  -ENOMEM;
* replaced `if (indio_dev == NULL`) with `if (!indio_dev)`
* used `strtobool()` instead of `kstrtou32()` inside the `ad5758_write_powerdown()` function
* used mutex_lock/unlock inside the `ad5758_write_powerdown()` function

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,784 @@
+	AD5758_DCDC_MODE_PPC_CURRENT,
+};
+
+static const int ad5758_min_max_table[][2] = {
```
the way you're doing this allows for invalid values to be selected, between 4 & 7 [via device-tree];
and there is that hole in the array between 4 & 7 which is undefined;
also, i don't really like arrays with holes in them [especially when they're tables of values]; not sure about other opinions ;

a more elegant way here, would be to have a struct of 
```
struct ad5758_range {
    int reg;
    int min;
    int max;
};
```
and then 
```
static const struct ad5758_range ad5758_min_max_table[] = {
      { 0,  0, 5000 },
       .....
};
```

this is preferable, because `ARRAY_SIZE(ad5758_min_max_table)` looks clearer than doing the same with a `int ad5758_min_max_table[][3] = ` definition

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,784 @@
+			dev_warn(&st->spi->dev,
+				"slew rate clock out of range, using default");
+
+		for (i = 0; i < ARRAY_SIZE(ad5758_slew_rate_step); i++) {
```
just curios: is there a generic search function for looking integers in an array of ints ?
in any case, it would be nice to wrap this code into a function here, since it looks like it's repeated for `ad5758_dc_dc_ilimt` and `ad5758_slew_rate_clk`

commodo commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,784 @@
+	return spi_write(st->spi, &st->data[0].d8[0], 4);
+}
+
+static int ad5758_parse_dt(struct ad5758_state *st)
```
if you're not using the return value, maybe make this a void return

stefpopa commented:
changelog v2 -> v3:
* used `static const struct ad5758_range ad5758_min_max_table[]` instead of `int ad5758_min_max_table[][3] `
* created `ad5758_get_array_index()` function which is used to look for valid integers in `ad5758_slew_rate_clk`, `ad5758_slew_rate_step` and `ad5758_dc_dc_ilimt` arrays.
* `ad5758_parse_dt `changed to void type

lclausen-adi commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,810 @@
+	if (ret < 0)
+		return ret;
+
+	return be16_to_cpu(st->data[2].d32 >> 8);
```
Does this work? Should this be be32_to_cpu(d32) >> 8? Maybe it works by accident because we either shift out the top 8 bits or the lower 8 bits depending on the endianness.  But still the tools will complain because you pass a __be32 to a function that expects a __be16.

Make sure you always build your drivers with `make C=2 CF=-D__CHECK_ENDIAN__`

lclausen-adi commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,810 @@
+	unsigned int regval;
+
+	regval = ad5758_spi_reg_read(st, addr);
+	if (regval < 0)
```
regval is unsigned. Never less than 0.

lclausen-adi commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,810 @@
+
+	timeout = 4;
+	do {
+		readval = ad5758_spi_reg_read(st, reg);
```
what if spi_reg_read() returns an error?

lclausen-adi commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,810 @@
+		ret =  ad5758_spi_reg_write(st, reg, writeval);
+	} else {
+		*readval = ad5758_spi_reg_read(st, reg);
+		if (*readval < 0)
```
readval is unsigned, never less than 0.

lclausen-adi commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,810 @@
+	return ad5758_spi_reg_write(st, addr, regval);
+}
+
+static int ad5758_get_array_index(const int *array, u8 size, int val)
```
why is size u8? That just makes the generated code more complicated.

lclausen-adi commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,810 @@
+				   unsigned int sr_clk,
+				   unsigned int sr_step)
+{
+	unsigned int msk, mode;
```
mask, I think we can spare the extra consonant ;)

lclausen-adi commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,810 @@
+{
+	struct ad5758_state *st = iio_priv(indio_dev);
+
+	return sprintf(buf, "%d\n", !!st->pwr_down);
```
I don't think we need the !!, we know pwr_down is either 0 or 1.

lclausen-adi commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,810 @@
+	return spi_write(st->spi, &st->data[0].d8[0], 4);
+}
+
+void ad5758_parse_dt(struct ad5758_state *st)
```
Kernel-Clippy would say: "It looks as if ad5758_parse_dt() is not exported. Do you want me to make it static?"

lclausen-adi commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,810 @@
+				    AD5758_DAC_CONFIG_OUT_EN_MSK,
+				    AD5758_DAC_CONFIG_OUT_EN_MODE(1));
+	if (ret < 0)
+		return ret;
```
You'll return ret here, no matter what. Might as well drop the check.

stefpopa commented:
changelog v3 -> v4:
* used `st->lock` instead of `indio_dev->mlock`
* protected the write_raw/read_raw and reg_access with mlock
* `ad5758_spi_reg_read()` returns `be32_to_cpu()` instead of `be16_to_cpu()`
* in several places the return of `ad5758_spi_reg_read() `was wrongly placed in an unsigned variable
* changed the size parameter type used in `ad5758_get_array_index()`
* msk variable name changed to mask
* removed the` !!` from `!!st->pwr_down`
* `ad5758_parse_dt `changed to static
* dropped the final` if (ret < 0)` check from `ad5758_init()`
* check if `ad5758_spi_reg_read()` returns error in `ad5758_wait_for_task_complete() `function

lclausen-adi commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,827 @@
+		return ret;
+
+	regval = ad5758_spi_reg_read(st, AD5758_DIGITAL_DIAG_RESULTS);
+	if (regval < 0)
```
Still unsigned

lclausen-adi commented on drivers/iio/dac/ad5758.c:
```diff
@@ -0,0 +1,827 @@
+			    struct iio_chan_spec const *chan,
+			    int val,
+			    int val2,
+			    long mask)
```
Not really important, but we should call this info instead of mask. You might also want to consider putting a few more things onto a single line here.

stefpopa commented:
changelog v4 -> v5:
* changed the type of regval in` ad5758_init()` from unsigned int to int
* renamed the `mask `parameter to `info `in `ad5758_write_raw() `function

stefpopa commented:
changelog v5 -> v6:
* fixed a few indentation inconsistencies 
* used `ad5758_output_range `enum elements as reg values for `ad5758_range `struct
